### DISCUSSION TITLE
How to include an R instruction as a parameter in a quarto document

### DISCUSSION THREAD

**Discussion by ddotta** [quarto-dev/quarto-cli#10991] [Q&A] [answered]
### Description

Hello !  

Is it possible to use an R instruction as a parameter in a quarto document?

With an .Rmd file, I used this syntax but it doesn't work :

```
params:
  com : !r c("01001", "01002", "01003")
```

```{r}
# params$com
```

I got this error 
```
ERROR: YAMLError: unknown tag !<!r> at line 6, column 43:
     ...  !r c("01001", "01002", "01003")
```

Many thanks in advance !

**Comment by mcanouil** [comment]
Could you edit the title of your post? Currently it is not meaningful.

See <https://quarto.org/docs/computations/r.html#chunk-options>.

**Reply by ddotta** [reply]
Done ! Sorry for the omission

**Comment by cderv** [comment, marked as answer]
> Is it possible to use an R instruction as a parameter in a quarto document?

No you can't do that anymore. You need to initialize the values using YAML syntax. It was possible in R Markdown because all happened in R, including the YAML parsing of options, whereas in Quarto R is used only for computation with **knitr**, and everything else does not use tools like R. This includes parsing parameters to pass them the engine, or allow auto completion, and YAML validation. 

If you really need to use R syntax, you can use the **quarto** R package to pass parameters, but this will be converted by the package to YAML equivalent. 

You can know what to write in YAML using **yaml** package

````
> cat(yaml::as.yaml(list(com = c("01001", "01002", "01003"))))
com:
- '01001'
- '01002'
- '01003'
````

**Reply by cderv** [reply]
To follow up on @mcanouil comment below, let me give more details 

`!r` and `!expr` are specific YAML tag handler defined in **knitr**. As I said, the YAML header in a .qmd is first handled by Quarto in all situation. We taught Quarto about `!expr` so that Quarto's YAML validation does not fail on this, as it could be handled by **knitr**. We did not do it for `!r`. So that is why you get an error message with `!r` and that `!expr` would work like shown below https://github.com/quarto-dev/quarto-cli/discussions/10991#discussioncomment-10866665

So technically, you could be using `!expr` inside your document instead of `!r`, but you can't use `!r` directly. However, as you are migrating, I would advice not using it when not needed and that is why I did not mention it. It is **knitr** specific and could be removed in the future (when dealing with a common interface for all engine https://github.com/quarto-dev/quarto-cli/issues/9197 and also because executing code in YAML is a bit of a security problem (`eval.expr = false` has been made the default in **yaml** R package now BTW)

Also if you are passing parameters at command line (which usually you would like to do), you will use [`--execute-param`](https://quarto.org/docs/computations/parameters.html#rendering) with a YAML file, and this one can't have `!expr` in it, because it will not be read by **knitr** but by Quarto. 

So overall, this is why I suggest not using `!expr` for params.

**Comment by ddotta** [comment]
Thanks for your great answer @cderv and @mcanouil !

**Comment by mcanouil** [comment]
See the following "knitr" example.
Note that Quarto discourages the use of `!expr` as it is not something working across engines (i.e., it is a knitr feature) and is not working everywhere.

````qmd
---
title: "Quarto Playground"
format: html
params:
  key: !expr "
    list(
      A = 1+1,
      B = letters
    )"
---

```{r}
params[["key"]][["A"]]
```
````


@ddotta Please edit the title to a meaningful one.

**Reply by al-obrien** [reply]
For people that aren't working in more than the knitr engine, I appreciate this is still possible via `!expr`, sometimes you want defaults set right within the header of the document and the R code is more natural for many users from Rmarkdown.
